### ISSUE TITLE
Matrix42 UEM Agent Windows

### ISSUE THREAD

**Issue by KUESTER-Group-Florian-Seip** [EUCweb/BIS-F#332]
The new Empirum Agent named "Matrix42 UEM Agent Windows" is not supported by BISF.

The UEM Agent no longer uses the "eris" service, but only the "Matrix42UAF" service and the AgentConfig.xml file is now located at C:\Program Files\Matrix42\Universal Agent Framework\AgentConfig.xml. So the script does not really clean up the agent configuration.

Some people will still use this agent because the vendor offers premium support for a special LTSR version.

Attached a few Screenshots
BISF Script Log
![2021-11-24 13_33_34-Window](https://user-images.githubusercontent.com/94964691/143239808-d1a78f68-539c-433d-b56b-358080865fb3.png)

BISF Script config servicename
![2021-11-24 13_34_22-Window](https://user-images.githubusercontent.com/94964691/143239815-e7179dcf-48d7-4948-b4f5-6ba3bc7615f2.png)

Matrix42 UEM Agent service name
![2021-11-24 13_35_28-Window](https://user-images.githubusercontent.com/94964691/143239816-dc4fa72c-3c52-4a75-931e-7ebd070f64b3.png)

**Comment by bladeoner** [comment]
@FSE-SYSADMIN can you test the attached script by replacing 10_PrepBISF_Empirum.ps1?
[10_PrepBISF_Empirum.zip](https://github.com/EUCweb/BIS-F/files/9332856/10_PrepBISF_Empirum.zip)

I've added the following:
Old: ```$ServiceNames = @("Eris", "MATRIXAUT")```
New: ```$ServiceNames = @("Eris", "MATRIXAUT", "Matrix42UAF")```

Thanks in advance.

If it works I can be send in as a PR.

**Comment by KUESTER-Group-Florian-Seip** [comment]
It doesn't work unfortunately. 
![image](https://user-images.githubusercontent.com/94964691/184804664-293c0188-30af-49e9-8942-dd959f1a2891.png)

I think that the variable $empirum_path must be changed to "C:\Program Files\Matrix42\Universal Agent Framework\". But as written in my previous post, the old agent should still be supported as well. We are currently in a transition phase. Unfortunately my Powershell skills are not yet sufficient for this ;)

**Comment by bladeoner** [comment]
@FSE-SYSADMIN thank you for the quick response.

I have an idea how to do it, I will look into it later this week and keep you informed.

**Comment by bladeoner** [comment]
@FSE-SYSADMIN I've added a ForEach, with that it can go through both locations.
[10_PrepBISF_Empirum_08182022.zip](https://github.com/EUCweb/BIS-F/files/9368297/10_PrepBISF_Empirum_08182022.zip)

Can you test if it works and let me know?

Thanks in advance.

**Comment by KUESTER-Group-Florian-Seip** [comment]
The same error in the BISF log file as above.

It seems that only the first service from the $servicename array is checked and since it is not present, the script exits.

**Comment by bladeoner** [comment]
@FSE-SYSADMIN I think I found the issue, on line 36 I changed the following:
Old: IF ($svc -eq $true) { Invoke-BISFService -ServiceName **"$($ServiceName)"** -Action Stop }
New: IF ($svc -eq $true) { Invoke-BISFService -ServiceName **"$ServiceName"** -Action Stop }

Could you try the attached file?
[10_PrepBISF_Empirum_08192022.zip](https://github.com/EUCweb/BIS-F/files/9380761/10_PrepBISF_Empirum_08192022.zip)

Thanks in advance.

**Comment by bladeoner** [comment]
I've replaced the function with the following:

	$svc = Test-BISFService -ServiceName "$Svc1"
	if ($svc -eq $true) {
		Invoke-BISFService -ServiceName "$Svc1" -Action Stop 
	}
	
	$svc = Test-BISFService -ServiceName "$Svc2"
	if ($svc -eq $true) {
		Invoke-BISFService -ServiceName "$Svc2" -Action Stop 
	}

	$svc = Test-BISFService -ServiceName "$Svc3"
	if ($svc -eq $true) {
		Invoke-BISFService -ServiceName "$Svc3" -Action Stop 
	}
Can you try the attached file and let me know if it works?
[10_PrepBISF_Empirum_08242022.zip](https://github.com/EUCweb/BIS-F/files/9414753/10_PrepBISF_Empirum_08242022.zip)

Thanks in advance.

I'm not able to test it completely, but when I run the modified file I get 3 responses in the logfile of all 3 services that they don't exist

**Comment by KUESTER-Group-Florian-Seip** [comment]
Now it works a little bit. The BISF log file is there now:
![image](https://user-images.githubusercontent.com/94964691/186582902-8a5d54a9-0a6c-4971-acf4-f78abc90e066.png)

But there are some errors in the PS output:
![image](https://user-images.githubusercontent.com/94964691/186582265-4657f019-2de6-4c4e-905d-ac3a57417667.png)

**Comment by bladeoner** [comment]
@FSE-SYSADMIN thank you for your patience and testing.

I made some adjustments on how to handle the errors.

When you find some time can you test the following file?
[10_PrepBISF_Empirum_08252022.zip](https://github.com/EUCweb/BIS-F/files/9421744/10_PrepBISF_Empirum_08252022.zip)

Thanks in advance.

**Comment by KUESTER-Group-Florian-Seip** [comment]
Yes, that looks much better! The first part of the removal (Agent LocalCache) works! Thank you very much for your work!

In the second part of the removal (reg keys) there are some reg keys that are not removed by the script:
		Remove-Item "$hklm_sw\MATRIX42\AGENT" -Force -ErrorAction SilentlyContinue
Has no change:
![image](https://user-images.githubusercontent.com/94964691/186587521-4a715e00-b7ad-4b84-85e1-abf90170f658.png)


		Remove-Item "$hklm_sw\MATRIX42\EmpInv" -Force -ErrorAction SilentlyContinue
Also has no change:
![image](https://user-images.githubusercontent.com/94964691/186587552-0e22c1f1-633e-4dc0-85d6-f09f2ef821d6.png)

**Comment by bladeoner** [comment]
Does it work when you run the following manually from a Powershell shell (admin)?
```Remove-Item "$hklm_sw\MATRIX42\AGENT" -Force```
```Remove-Item "$hklm_sw\MATRIX42\EmpInv" -Force```

**Comment by KUESTER-Group-Florian-Seip** [comment]
No. I get the error:
![image](https://user-images.githubusercontent.com/94964691/186589449-65b184b8-c6a9-46eb-961d-38ec8e83d3ac.png)

It seems that the variable $hklm_sw has no content.

**Comment by bladeoner** [comment]
I was testing it at the same time with another registry value on my machine, I also get an error.

Can you verify the following works?
```Remove-Item "HKLM:\Software\MATRIX42\AGENT" -Force```
```Remove-Item "HKLM:\Software\MATRIX42\EmpInv" -Force```

**Comment by bladeoner** [comment]
If it works attached is a new file containing the changes:
[10_PrepBISF_Empirum_08252022.zip](https://github.com/EUCweb/BIS-F/files/9421997/10_PrepBISF_Empirum_08252022.zip)

**Comment by KUESTER-Group-Florian-Seip** [comment]
Oh, this change needs to be confirmed:
![image](https://user-images.githubusercontent.com/94964691/186595389-97a763ac-0ef8-42e3-bb77-0d7cddf7acb1.png)

**Comment by bladeoner** [comment]
@FSE-SYSADMIN I've added the -Recurse parameter to the cleanup of the registry keys.
[10_PrepBISF_Empirum_08252022.zip](https://github.com/EUCweb/BIS-F/files/9422914/10_PrepBISF_Empirum_08252022.zip)

**Comment by KUESTER-Group-Florian-Seip** [comment]
YES. Now it works!
Thank you very much again!

**Comment by bladeoner** [comment]
That's good to hear, I will open a PR.
